File: tools/core-v02-node-verifier/lib/manifest.mjs
import { createHash } from "node:crypto";
import { lstat, readFile, readdir } from "node:fs/promises";
import path from "node:path";

export class ManifestError extends Error {
  constructor(message) {
    super(message);
    this.name = "ManifestError";
  }
}

const TOP_LEVEL_KEYS = new Set([
  "alpn",
  "format_version",
  "hash",
  "protocol",
  "protocol_version",
  "scenarios",
  "vectors",
]);
const VECTOR_KEYS = new Set([
  "artifact_path",
  "artifact_type",
  "class",
  "expected_error",
  "expected_outcome",
  "id",
  "length",
  "message_type",
  "scenario_only",
  "sha256",
  "validation_stage",
]);
const SCENARIO_KEYS = new Set(["final_assertions", "id", "initial_core_version", "steps"]);
const STEP_KEYS = new Set(["expected_error", "expected_outcome", "sequence", "vector_id"]);
const CLASSES = new Set(["valid", "invalid"]);
const OUTCOMES = new Set(["accept", "reject", "close"]);
const ARTIFACT_TYPES = new Set([
  "core-object-cbor",
  "cose-sign1",
  "control-envelope-cbor",
  "proof-transcript-cbor",
  "ed25519-signature",
  "malformed-bytes",
]);
const STAGES = new Set([
  "structural",
  "deterministic-cbor",
  "object-schema",
  "cose",
  "envelope-schema",
  "message-schema",
  "version-dispatch",
  "proof",
  "binding",
  "replay",
  "policy",
]);
const MESSAGE_TYPES = new Set([
  "CLIENT_HELLO",
  "EDGE_HELLO",
  "ROUTE_OPEN",
  "ROUTE_ACCEPT",
  "ROUTE_REJECT",
  "STREAM_OPEN",
  "STREAM_ACCEPT",
  "STREAM_REJECT",
  "LEASE_RENEW",
  "LEASE_RESULT",
  "KEY_UPDATE_NOTICE",
  "ROUTE_DRAIN",
  "ROUTE_REVOKE",
  "ROUTE_CLOSE",
  "PING",
  "PONG",
  "ERROR",
]);
const ERROR_NAMES = new Set([
  "NBSR_E_NAME_INVALID",
  "NBSR_E_NAME_NOT_FOUND",
  "NBSR_E_RECORD_UNTRUSTED",
  "NBSR_E_RECORD_STALE",
  "NBSR_E_RECORD_REVOKED",
  "NBSR_E_CONTEXT_REQUIRED",
  "NBSR_E_HANDLE_EXHAUSTED",
  "NBSR_E_ROUTE_DENIED",
  "NBSR_E_GRANT_INVALID",
  "NBSR_E_GRANT_EXPIRED",
  "NBSR_E_PROOF_INVALID",
  "NBSR_E_REPLAY",
  "NBSR_E_PROFILE_UNSUPPORTED",
  "NBSR_E_DOWNGRADE",
  "NBSR_E_EDGE_UNAVAILABLE",
  "NBSR_E_ORIGIN_UNAVAILABLE",
  "NBSR_E_REVOKED",
  "NBSR_E_OVER_CAPACITY",
  "NBSR_E_INTERNAL",
]);
const FINAL_ASSERTIONS = new Set([
  "transport-session-active",
  "transport-session-closed",
  "route-context-active",
  "no-route-state",
  "stream-active",
  "no-stream-state",
  "replay-state-unchanged",
  "replay-tombstone-retained",
  "no-version-fallback",
  "no-origin-disclosure",
]);
const SUPPORT_FILES = new Set([
  "README.md",
  "manifest.json",
  "keys/test-only-route-grant-ed25519-seed.hex",
  "keys/test-only-route-grant-ed25519-public.hex",
  "keys/test-only-session-ed25519-seed.hex",
  "keys/test-only-session-ed25519-public.hex",
]);
const ID_PATTERN = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const SHA256_PATTERN = /^[0-9a-f]{64}$/;

function fail(message) {
  throw new ManifestError(message);
}

function scanJson(text) {
  let offset = 0;

  function whitespace() {
    while (offset < text.length && /\s/u.test(text[offset])) {
      offset += 1;
    }
  }

  function string() {
    const start = offset;
    if (text[offset] !== '"') {
      fail("invalid manifest JSON");
    }
    offset += 1;
    while (offset < text.length) {
      const character = text[offset];
      if (character === '"') {
        offset += 1;
        try {
          return JSON.parse(text.slice(start, offset));
        } catch {
          fail("invalid manifest JSON");
        }
      }
      if (character === "\\") {
        offset += 1;
        if (offset >= text.length) {
          fail("invalid manifest JSON");
        }
        if (text[offset] === "u") {
          if (!/^[0-9a-fA-F]{4}$/u.test(text.slice(offset + 1, offset + 5))) {
            fail("invalid manifest JSON");
          }
          offset += 5;
          continue;
        }
        if (!'"\\/bfnrt'.includes(text[offset])) {
          fail("invalid manifest JSON");
        }
      } else if (character.charCodeAt(0) < 0x20) {
        fail("invalid manifest JSON");
      }
      offset += 1;
    }
    fail("invalid manifest JSON");
  }

  function value() {
    whitespace();
    const character = text[offset];
    if (character === "{") {
      object();
      return;
    }
    if (character === "[") {
      array();
      return;
    }
    if (character === '"') {
      string();
      return;
    }
    const remainder = text.slice(offset);
    const token = remainder.match(/^(?:true|false|null|-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?)/u);
    if (!token) {
      fail("invalid manifest JSON");
    }
    offset += token[0].length;
  }

  function object() {
    offset += 1;
    whitespace();
    const keys = new Set();
    if (text[offset] === "}") {
      offset += 1;
      return;
    }
    while (offset < text.length) {
      whitespace();
      const key = string();
      if (keys.has(key)) {
        fail(`duplicate JSON key: ${key}`);
      }
      keys.add(key);
      whitespace();
      if (text[offset] !== ":") {
        fail("invalid manifest JSON");
      }
      offset += 1;
      value();
      whitespace();
      if (text[offset] === "}") {
        offset += 1;
        return;
      }
      if (text[offset] !== ",") {
        fail("invalid manifest JSON");
      }
      offset += 1;
    }
    fail("invalid manifest JSON");
  }

  function array() {
    offset += 1;
    whitespace();
    if (text[offset] === "]") {
      offset += 1;
      return;
    }
    while (offset < text.length) {
      value();
      whitespace();
      if (text[offset] === "]") {
        offset += 1;
        return;
      }
      if (text[offset] !== ",") {
        fail("invalid manifest JSON");
      }
      offset += 1;
    }
    fail("invalid manifest JSON");
  }

  value();
  whitespace();
  if (offset !== text.length) {
    fail("invalid manifest JSON");
  }
}

export function parseJsonStrict(text) {
  if (typeof text !== "string" || text.startsWith("\uFEFF")) {
    fail("invalid manifest JSON");
  }
  scanJson(text);
  try {
    return JSON.parse(text);
  } catch {
    fail("invalid manifest JSON");
  }
}

function requireExactKeys(value, expected, label) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    fail(`${label} must be an object`);
  }
  const actual = Object.keys(value);
  if (actual.length !== expected.size || actual.some((key) => !expected.has(key))) {
    fail(`${label} has unknown or missing keys`);
  }
}

function requireInteger(value, label, minimum = 0) {
  if (!Number.isSafeInteger(value) || value < minimum) {
    fail(`invalid ${label}`);
  }
}

function requireId(value, label) {
  if (typeof value !== "string" || !ID_PATTERN.test(value)) {
    fail(`invalid ${label}`);
  }
}

function validateOutcome(outcome, error) {
  if (!OUTCOMES.has(outcome)) {
    fail("invalid expected_outcome");
  }
  if (outcome === "reject") {
    if (!ERROR_NAMES.has(error)) {
      fail("reject must name a frozen error");
    }
  } else if (error !== null) {
    fail(`${outcome} expected_error must be null`);
  }
}

function validateArtifactPath(artifactPath) {
  if (
    typeof artifactPath !== "string"
    || artifactPath.includes("\\")
    || artifactPath.includes("://")
    || /^[A-Za-z]:/u.test(artifactPath)
    || artifactPath.startsWith("/")
  ) {
    fail("invalid artifact path");
  }
  const parts = artifactPath.split("/");
  if (parts[0] !== "artifacts" || parts.some((part) => !part || part === "." || part === "..")) {
    fail("invalid artifact path");
  }
}

function validateVector(vector) {
  requireExactKeys(vector, VECTOR_KEYS, "vector");
  requireId(vector.id, "vector id");
  if (!CLASSES.has(vector.class)) {
    fail("invalid class");
  }
  if (!ARTIFACT_TYPES.has(vector.artifact_type)) {
    fail("invalid artifact_type");
  }
  validateArtifactPath(vector.artifact_path);
  requireInteger(vector.length, "length");
  if (typeof vector.sha256 !== "string" || !SHA256_PATTERN.test(vector.sha256)) {
    fail("invalid sha256");
  }
  validateOutcome(vector.expected_outcome, vector.expected_error);
  if (!STAGES.has(vector.validation_stage)) {
    fail("invalid validation_stage");
  }
  if (vector.expected_outcome === "close" && !["structural", "version-dispatch"].includes(vector.validation_stage)) {
    fail("close requires structural or version-dispatch stage");
  }
  if (vector.message_type !== null && !MESSAGE_TYPES.has(vector.message_type)) {
    fail("invalid message_type");
  }
  if (typeof vector.scenario_only !== "boolean") {
    fail("invalid scenario_only");
  }
}

function validateScenario(scenario, vectorIds) {
  requireExactKeys(scenario, SCENARIO_KEYS, "scenario");
  requireId(scenario.id, "scenario id");
  requireInteger(scenario.initial_core_version, "initial_core_version", 1);
  if (!Array.isArray(scenario.steps) || scenario.steps.length === 0) {
    fail("scenario steps must be nonempty");
  }
  for (const [index, step] of scenario.steps.entries()) {
    requireExactKeys(step, STEP_KEYS, "scenario step");
    requireInteger(step.sequence, "scenario sequence", 1);
    if (step.sequence !== index + 1) {
      fail("scenario sequence must be contiguous");
    }
    requireId(step.vector_id, "scenario vector id");
    if (!vectorIds.has(step.vector_id)) {
      fail("unknown scenario vector id");
    }
    validateOutcome(step.expected_outcome, step.expected_error);
  }
  if (
    !Array.isArray(scenario.final_assertions)
    || new Set(scenario.final_assertions).size !== scenario.final_assertions.length
    || scenario.final_assertions.some((assertion) => !FINAL_ASSERTIONS.has(assertion))
  ) {
    fail("invalid final_assertions");
  }
}

export function validateManifest(manifest) {
  requireExactKeys(manifest, TOP_LEVEL_KEYS, "manifest");
  if (manifest.format_version !== 1 || !Number.isInteger(manifest.format_version)) {
    fail("invalid format_version");
  }
  if (manifest.protocol !== "NBSR") {
    fail("invalid protocol");
  }
  if (manifest.protocol_version !== 2 || !Number.isInteger(manifest.protocol_version)) {
    fail("invalid protocol_version");
  }
  if (manifest.alpn !== "nbsr-quic-1" || manifest.hash !== "sha256") {
    fail("invalid protocol profile");
  }
  if (!Array.isArray(manifest.vectors) || !Array.isArray(manifest.scenarios)) {
    fail("vectors and scenarios must be arrays");
  }

  for (const vector of manifest.vectors) {
    validateVector(vector);
  }
  const vectorIds = manifest.vectors.map((vector) => vector.id);
  const artifactPaths = manifest.vectors.map((vector) => vector.artifact_path);
  if (new Set(vectorIds).size !== vectorIds.length || new Set(artifactPaths).size !== artifactPaths.length) {
    fail("duplicate vector id or artifact path");
  }
  if (vectorIds.some((id, index) => index > 0 && id <= vectorIds[index - 1])) {
    fail("vectors must be sorted by id");
  }

  const knownVectorIds = new Set(vectorIds);
  for (const scenario of manifest.scenarios) {
    validateScenario(scenario, knownVectorIds);
  }
  const scenarioIds = manifest.scenarios.map((scenario) => scenario.id);
  if (
    new Set(scenarioIds).size !== scenarioIds.length
    || scenarioIds.some((id, index) => index > 0 && id <= scenarioIds[index - 1])
  ) {
    fail("scenarios must have unique sorted ids");
  }
  return manifest;
}

export async function loadManifest(vectorRoot) {
  try {
    const manifestPath = path.join(path.resolve(vectorRoot), "manifest.json");
    const text = await readFile(manifestPath, "utf8");
    return validateManifest(parseJsonStrict(text));
  } catch (error) {
    if (error instanceof ManifestError) {
      throw error;
    }
    throw new ManifestError("invalid manifest JSON");
  }
}

async function listRegularFiles(root, current = root) {
  const relativeFiles = [];
  let entries;
  try {
    entries = await readdir(current, { withFileTypes: true });
  } catch {
    fail("invalid vector package");
  }
  for (const entry of entries) {
    const absolute = path.join(current, entry.name);
    const relative = path.relative(root, absolute).split(path.sep).join("/");
    if (entry.isSymbolicLink()) {
      fail(`symbolic link prohibited: ${relative}`);
    }
    if (entry.isDirectory()) {
      relativeFiles.push(...await listRegularFiles(root, absolute));
    } else if (entry.isFile()) {
      relativeFiles.push(relative);
    } else {
      fail(`unsupported package entry: ${relative}`);
    }
  }
  return relativeFiles;
}

export async function verifyPackageInventory(vectorRoot, manifest) {
  validateManifest(manifest);
  const root = path.resolve(vectorRoot);
  let rootStat;
  try {
    rootStat = await lstat(root);
  } catch {
    fail("invalid vector package");
  }
  if (!rootStat.isDirectory() || rootStat.isSymbolicLink()) {
    fail("invalid vector package");
  }

  const expectedArtifacts = new Set(manifest.vectors.map((vector) => vector.artifact_path));
  const expectedFiles = new Set([...SUPPORT_FILES, ...expectedArtifacts]);
  const actualFiles = new Set(await listRegularFiles(root));
  if (
    expectedFiles.size !== actualFiles.size
    || [...expectedFiles].some((file) => !actualFiles.has(file))
  ) {
    fail("package file inventory mismatch");
  }

  const artifacts = new Map();
  for (const vector of manifest.vectors) {
    const absolute = path.resolve(root, ...vector.artifact_path.split("/"));
    if (!absolute.startsWith(`${root}${path.sep}`)) {
      fail("artifact path escapes package");
    }
    const stat = await lstat(absolute);
    if (!stat.isFile() || stat.isSymbolicLink()) {
      fail(`invalid artifact: ${vector.id}`);
    }
    const data = await readFile(absolute);
    if (data.length !== vector.length) {
      fail(`artifact length mismatch: ${vector.id}`);
    }
    if (createHash("sha256").update(data).digest("hex") !== vector.sha256) {
      fail(`artifact SHA-256 mismatch: ${vector.id}`);
    }
    artifacts.set(vector.id, data);
  }
  return artifacts;
}

File: scripts/core_v02_vectors/manifest.py
from __future__ import annotations

import json
import re
from dataclasses import asdict, dataclass
from hashlib import sha256
from pathlib import Path, PurePosixPath
from typing import Any

from nbsr.protocol.registry import ErrorCode, MessageType


VECTOR_ID = re.compile(r"[a-z0-9]+(?:-[a-z0-9]+)*\Z")
SHA256_HEX = re.compile(r"[0-9a-f]{64}\Z")

VECTOR_CLASSES = frozenset({"valid", "invalid"})
ARTIFACT_TYPES = frozenset(
    {
        "core-object-cbor",
        "cose-sign1",
        "control-envelope-cbor",
        "proof-transcript-cbor",
        "ed25519-signature",
        "malformed-bytes",
    }
)
OUTCOMES = frozenset({"accept", "reject"})
VALIDATION_STAGES = frozenset(
    {
        "structural",
        "deterministic-cbor",
        "object-schema",
        "cose",
        "envelope-schema",
        "message-schema",
        "version-dispatch",
        "proof",
        "binding",
        "replay",
        "policy",
    }
)
FINAL_ASSERTIONS = frozenset(
    {
        "transport-session-active",
        "transport-session-closed",
        "route-context-active",
        "no-route-state",
        "stream-active",
        "no-stream-state",
        "replay-state-unchanged",
        "replay-tombstone-retained",
        "no-version-fallback",
        "no-origin-disclosure",
    }
)
ERROR_NAMES = frozenset(item.name for item in ErrorCode)
MESSAGE_NAMES = frozenset(item.name for item in MessageType)

TOP_LEVEL_KEYS = frozenset(
    {
        "format_version",
        "protocol",
        "protocol_version",
        "alpn",
        "hash",
        "vectors",
        "scenarios",
    }
)
VECTOR_KEYS = frozenset(
    {
        "id",
        "class",
        "artifact_type",
        "artifact_path",
        "length",
        "sha256",
        "expected_outcome",
        "expected_error",
        "validation_stage",
        "message_type",
        "scenario_only",
    }
)
SCENARIO_KEYS = frozenset({"id", "initial_core_version", "steps", "final_assertions"})
STEP_KEYS = frozenset({"sequence", "vector_id", "expected_outcome", "expected_error"})
SUPPORT_FILES = frozenset(
    {
        "keys/test-only-route-grant-ed25519-seed.hex",
        "keys/test-only-route-grant-ed25519-public.hex",
        "keys/test-only-session-ed25519-seed.hex",
        "keys/test-only-session-ed25519-public.hex",
    }
)


@dataclass(frozen=True, slots=True)
class VectorEntry:
    id: str
    vector_class: str
    artifact_type: str
    artifact_path: str
    length: int
    sha256: str
    expected_outcome: str
    expected_error: str | None
    validation_stage: str
    message_type: str | None
    scenario_only: bool


@dataclass(frozen=True, slots=True)
class ScenarioStep:
    sequence: int
    vector_id: str
    expected_outcome: str
    expected_error: str | None


@dataclass(frozen=True, slots=True)
class ScenarioEntry:
    id: str
    initial_core_version: int
    steps: tuple[ScenarioStep, ...]
    final_assertions: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class VectorManifest:
    format_version: int
    protocol: str
    protocol_version: int
    alpn: str
    hash: str
    vectors: tuple[VectorEntry, ...]
    scenarios: tuple[ScenarioEntry, ...]


def _reject_unknown_keys(value: dict[str, object], expected: frozenset[str]) -> None:
    unknown = set(value) - expected
    missing = expected - set(value)
    if unknown:
        raise ValueError(f"unknown manifest keys: {sorted(unknown)}")
    if missing:
        raise ValueError(f"missing manifest keys: {sorted(missing)}")


def _require_int(value: object, name: str, *, minimum: int = 0) -> int:
    if type(value) is not int or value < minimum:
        raise ValueError(f"invalid {name}")
    return value


def _require_string(value: object, name: str) -> str:
    if not isinstance(value, str):
        raise ValueError(f"invalid {name}")
    return value


def _require_id(value: object, name: str = "id") -> str:
    result = _require_string(value, name)
    if VECTOR_ID.fullmatch(result) is None:
        raise ValueError(f"invalid {name}")
    return result


def _validate_outcome(outcome: object, error: object) -> tuple[str, str | None]:
    valid_outcome = _require_string(outcome, "expected_outcome")
    if valid_outcome not in OUTCOMES:
        raise ValueError("invalid expected_outcome")
    if valid_outcome == "accept":
        if error is not None:
            raise ValueError("expected_error must be null for accept")
        return valid_outcome, None
    if not isinstance(error, str) or error not in ERROR_NAMES:
        raise ValueError("expected_error must name a frozen error")
    return valid_outcome, error


def _validate_artifact_path(value: object) -> str:
    path = _require_string(value, "artifact_path")
    if "\\" in path or "://" in path or re.match(r"^[A-Za-z]:", path) or path.startswith("/"):
        raise ValueError("invalid artifact path")
    parts = PurePosixPath(path).parts
    if not parts or any(part in {"", ".", ".."} for part in parts):
        raise ValueError("invalid artifact path")
    if parts[0] != "artifacts":
        raise ValueError("artifact path must be below artifacts")
    return path


def _parse_vector(raw: object) -> VectorEntry:
    if not isinstance(raw, dict):
        raise ValueError("vector entry must be an object")
    _reject_unknown_keys(raw, VECTOR_KEYS)
    vector_class = _require_string(raw["class"], "class")
    if vector_class not in VECTOR_CLASSES:
        raise ValueError("invalid class")
    artifact_type = _require_string(raw["artifact_type"], "artifact_type")
    if artifact_type not in ARTIFACT_TYPES:
        raise ValueError("invalid artifact_type")
    checksum = _require_string(raw["sha256"], "sha256")
    if SHA256_HEX.fullmatch(checksum) is None:
        raise ValueError("invalid sha256")
    stage = _require_string(raw["validation_stage"], "validation_stage")
    if stage not in VALIDATION_STAGES:
        raise ValueError("invalid validation_stage")
    message_type = raw["message_type"]
    if message_type is not None and (not isinstance(message_type, str) or message_type not in MESSAGE_NAMES):
        raise ValueError("invalid message_type")
    scenario_only = raw["scenario_only"]
    if type(scenario_only) is not bool:
        raise ValueError("invalid scenario_only")
    outcome, error = _validate_outcome(
        raw["expected_outcome"],
        raw["expected_error"],
    )
    return VectorEntry(
        id=_require_id(raw["id"]),
        vector_class=vector_class,
        artifact_type=artifact_type,
        artifact_path=_validate_artifact_path(raw["artifact_path"]),
        length=_require_int(raw["length"], "length"),
        sha256=checksum,
        expected_outcome=outcome,
        expected_error=error,
        validation_stage=stage,
        message_type=message_type,
        scenario_only=scenario_only,
    )


def _parse_step(raw: object) -> ScenarioStep:
    if not isinstance(raw, dict):
        raise ValueError("scenario step must be an object")
    _reject_unknown_keys(raw, STEP_KEYS)
    outcome, error = _validate_outcome(
        raw["expected_outcome"],
        raw["expected_error"],
    )
    return ScenarioStep(
        sequence=_require_int(raw["sequence"], "sequence", minimum=1),
        vector_id=_require_id(raw["vector_id"], "vector_id"),
        expected_outcome=outcome,
        expected_error=error,
    )


def _parse_scenario(raw: object) -> ScenarioEntry:
    if not isinstance(raw, dict):
        raise ValueError("scenario entry must be an object")
    _reject_unknown_keys(raw, SCENARIO_KEYS)
    raw_steps = raw["steps"]
    raw_assertions = raw["final_assertions"]
    if not isinstance(raw_steps, list) or not raw_steps:
        raise ValueError("scenario steps must be a nonempty array")
    if not isinstance(raw_assertions, list):
        raise ValueError("final_assertions must be an array")
    steps = tuple(_parse_step(step) for step in raw_steps)
    if tuple(step.sequence for step in steps) != tuple(range(1, len(steps) + 1)):
        raise ValueError("scenario sequence must be contiguous from one")
    assertions = tuple(_require_string(value, "final_assertion") for value in raw_assertions)
    if len(set(assertions)) != len(assertions) or any(value not in FINAL_ASSERTIONS for value in assertions):
        raise ValueError("invalid final_assertions")
    return ScenarioEntry(
        id=_require_id(raw["id"]),
        initial_core_version=_require_int(
            raw["initial_core_version"],
            "initial_core_version",
            minimum=1,
        ),
        steps=steps,
        final_assertions=assertions,
    )


def _reject_duplicate_pairs(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            raise ValueError(f"duplicate JSON key: {key}")
        result[key] = value
    return result


def _parse_manifest(raw: object) -> VectorManifest:
    if not isinstance(raw, dict):
        raise ValueError("manifest must be an object")
    _reject_unknown_keys(raw, TOP_LEVEL_KEYS)
    if raw["format_version"] != 1 or type(raw["format_version"]) is not int:
        raise ValueError("invalid format_version")
    if raw["protocol"] != "NBSR":
        raise ValueError("invalid protocol")
    if raw["protocol_version"] != 2 or type(raw["protocol_version"]) is not int:
        raise ValueError("invalid protocol_version")
    if raw["alpn"] != "nbsr-quic-1":
        raise ValueError("invalid alpn")
    if raw["hash"] != "sha256":
        raise ValueError("invalid hash")
    if not isinstance(raw["vectors"], list) or not isinstance(raw["scenarios"], list):
        raise ValueError("vectors and scenarios must be arrays")
    vectors = tuple(_parse_vector(item) for item in raw["vectors"])
    scenarios = tuple(_parse_scenario(item) for item in raw["scenarios"])
    vector_ids = [item.id for item in vectors]
    artifact_paths = [item.artifact_path for item in vectors]
    scenario_ids = [item.id for item in scenarios]
    if len(set(vector_ids)) != len(vector_ids):
        raise ValueError("duplicate vector id")
    if len(set(artifact_paths)) != len(artifact_paths):
        raise ValueError("duplicate artifact path")
    if len(set(scenario_ids)) != len(scenario_ids):
        raise ValueError("duplicate scenario id")
    known_vectors = set(vector_ids)
    for scenario in scenarios:
        for step in scenario.steps:
            if step.vector_id not in known_vectors:
                raise ValueError("unknown scenario vector_id")
    return VectorManifest(
        format_version=1,
        protocol="NBSR",
        protocol_version=2,
        alpn="nbsr-quic-1",
        hash="sha256",
        vectors=vectors,
        scenarios=scenarios,
    )


def load_manifest(path: Path) -> VectorManifest:
    try:
        text = path.read_text(encoding="utf-8-sig")
        raw = json.loads(text, object_pairs_hook=_reject_duplicate_pairs)
    except (OSError, UnicodeError, json.JSONDecodeError) as exc:
        raise ValueError("invalid manifest JSON") from exc
    return _parse_manifest(raw)


def _manifest_dict(manifest: VectorManifest) -> dict[str, object]:
    vectors = []
    for entry in sorted(manifest.vectors, key=lambda item: item.id):
        value = asdict(entry)
        value["class"] = value.pop("vector_class")
        vectors.append(value)
    scenarios = [asdict(item) for item in sorted(manifest.scenarios, key=lambda item: item.id)]
    return {
        "format_version": manifest.format_version,
        "protocol": manifest.protocol,
        "protocol_version": manifest.protocol_version,
        "alpn": manifest.alpn,
        "hash": manifest.hash,
        "vectors": vectors,
        "scenarios": scenarios,
    }


def dump_manifest(manifest: VectorManifest) -> bytes:
    validated = _parse_manifest(_manifest_dict(manifest))
    return (
        json.dumps(
            _manifest_dict(validated),
            indent=2,
            sort_keys=True,
            ensure_ascii=False,
        )
        + "\n"
    ).encode("utf-8")


def validate_package(root: Path, manifest: VectorManifest) -> None:
    resolved_root = root.resolve()
    expected = {entry.artifact_path for entry in manifest.vectors}
    for entry in manifest.vectors:
        artifact = (root / Path(*PurePosixPath(entry.artifact_path).parts)).resolve()
        if not artifact.is_relative_to(resolved_root):
            raise ValueError(f"artifact path escapes package: {entry.artifact_path}")
        if not artifact.is_file():
            raise ValueError(f"missing artifact: {entry.artifact_path}")
        payload = artifact.read_bytes()
        if len(payload) != entry.length:
            raise ValueError(f"length mismatch: {entry.artifact_path}")
        if sha256(payload).hexdigest() != entry.sha256:
            raise ValueError(f"sha mismatch: {entry.artifact_path}")
    actual: set[str] = set()
    for directory in ("artifacts", "keys"):
        base = root / directory
        if base.exists():
            actual.update(path.relative_to(root).as_posix() for path in base.rglob("*") if path.is_file())
    extras = actual - expected - SUPPORT_FILES
    if extras:
        raise ValueError(f"extra package files: {sorted(extras)}")
